aptos-move/aptos-aggregator/src/bounded_math.rs

```diff
@@ -0,0 +1,214 @@
+    match value {
+        Ok(value) => Ok(Some(value)),
+        Err(BoundedMathError::Overflow) => Err(BoundedMathError::Overflow),
+        Err(BoundedMathError::Underflow) => Ok(None),
+    }
+}
+
+pub fn expect_ok<T>(value: BoundedMathResult<T>) -> PartialVMResult<T> {
+    value.map_err(|e| {
+        error!("Aggregator code invariant broken (there is a bug in the code)");
+        abort_error(
+            format!(
+                "Aggregator code invariant broken (there is a bug in the code), {:?}",
+                e
+            ),
+            ECODE_INVARIANT_BROKEN,
+        )
+    })
+}
+
+/// Implements application of `Addition` to `base`.
+pub fn addition(base: u128, value: u128, limit: u128) -> BoundedMathResult<u128> {
+    if limit < base || value > (limit - base) {
+        Err(BoundedMathError::Overflow)
+    } else {
+        Ok(base + value)
+    }
+}
+
+/// Implements application of `Subtraction` to `base`.
+pub fn subtraction(base: u128, value: u128) -> BoundedMathResult<u128> {
+    if value > base {
+        Err(BoundedMathError::Underflow)
+    } else {
+        Ok(base - value)
+    }
+}
+
+/// Describes the delta of an aggregator.
+/// Rename to SignedU128 ?
+#[derive(Clone, Copy, Hash, PartialOrd, Ord, Debug, PartialEq, Eq)]
+pub enum DeltaValue {
+    Positive(u128),
+    Negative(u128),
+}
+
+impl DeltaValue {
+    pub fn minus(&self) -> Self {
+        match self {
+            DeltaValue::Positive(value) => DeltaValue::Negative(*value),
+            DeltaValue::Negative(value) => DeltaValue::Positive(*value),
+        }
+    }
+
+    pub fn add(&self, other: &Self, max_value: u128) -> BoundedMathResult<Self> {
+        // Another useful macro, this time for merging deltas with different signs, such
+        // as +A-B and -A+B. In these cases we have to check which of A or B is greater
+        // and possibly flip a sign.
+        macro_rules! update_different_sign {
+            ($a:ident, $b:ident) => {
+                if $a >= $b {
+                    DeltaValue::Positive(subtraction(*$a, *$b)?)
+                } else {
+                    DeltaValue::Negative(subtraction(*$b, *$a)?)
+                }
+            };
+        }
+
+        Ok(match (self, other) {
+            (DeltaValue::Positive(v1), DeltaValue::Positive(v2)) => {
+                DeltaValue::Positive(addition(*v1, *v2, max_value)?)
+            },
+            (DeltaValue::Positive(v1), DeltaValue::Negative(v2)) => update_different_sign!(v1, v2),
+            (DeltaValue::Negative(v1), DeltaValue::Positive(v2)) => update_different_sign!(v2, v1),
+            (DeltaValue::Negative(v1), DeltaValue::Negative(v2)) => {
+                DeltaValue::Negative(addition(*v1, *v2, max_value).map_err(|err| {
+                    if err == BoundedMathError::Overflow {
+                        BoundedMathError::Underflow
+                    } else {
+                        err
+                    }
+                })?)
+            },
```

Comment:
let's create a function to negate_error(BoundedMathResult), and invert the other side too, i.e:

```
fn negate_error<T>(result: BoundedMathResult<T>): BoundedMathResult<T> {
  result.map_err(|err| {
    match err {
      BoundedMathError::Overflow => BoundedMathError::Underflow,
      BoundedMathError::Underflow => BoundedMathError::Overflow,
    }
  }
}

```